common/src/Polyhedron.h

```diff
@@ -23,351 +23,1333 @@
+     *
+     * If no vertex is within the given max distance of the given points, then the given max distance is
+     * returned.
+     *
+     * Then it proceeds to compute the distance of the succeeding vertices (in the order in which the vertices
+     * appear in the boundary) to the succeeding given positions, and takes the maximum of all computed
+     * distances and returns it.
+     *
+     * @param positions the positions to compute the distance to
+     * @param maxDistance the max distance
+     * @return the maximum distance as computed
+     */
+    T distanceTo(const std::vector<vm::vec<T,3>>& positions, T maxDistance = std::numeric_limits<T>::max()) const;
+
+    /**
+     * Returns the normal of this face. The normal is computed by finding three non colinear vertices in the
+     * boundary of this face, and taking the cross product of the vectors between these vertices.
+     *
+     * If no colinear vertices are found, then a zero vector is returned.
+     *
+     * @return the normal of this face
+     */
+    vm::vec<T,3> normal() const;
+
+    /**
+     * Returns the center of this face.
+     */
+    vm::vec<T,3> center() const;
+
+    /**
+     * Intersects this face with the given ray and returns the distance along the given ray's direction, from
+     * the given ray's origin to the point of intersection, or NaN if the given reay does not intersect this
+     * face.
+     *
+     * The given side determines whether hits from above or below this face should be returned. A hit is
+     * considered from above if the ray's origin is above this face (the face normal points towards it) and
+     * it is considered from below if the ray's origin is below this face (the face normal points away from
+     * it).
+     *
+     * @param ray the ray
+     * @param side the side(s) from which a hit should be considered
+     * @return the distance to the point of intersection or NaN if the given ray does not intersect this face
+     * from the given side
+     */
+    T intersectWithRay(const vm::ray<T,3>& ray, const vm::side side) const;
+
+    /**
+     * Computes the position of the given point in relation to the plane on which this face lies. This plane
+     * is determined by the position of the origin of the first half edge of this face and the face normal.
+     *
+     * @param point the point to check
+     * @param epsilon the epsilon value to use for the position check
+     * @return the relative position of the given point
+     */
+    vm::plane_status pointStatus(const vm::vec<T,3>& point, T epsilon = vm::constants<T>::point_status_epsilon()) const;
+
+    /**
+     * Prints a textual description of the face to the given output stream.
+     *
+     * @param stream the stream
+     * @param edge the edge to print
+     * @return a reference to the given stream
+     */
+    friend std::ostream& operator<<(std::ostream& stream, const Face& face) {
```

Comment:
Can't these be in the "implementation" header?

Reply:
These are free template functions, not member functions, so they'd have to be explicitly instantiated like the class templates. I didn't think this would make much of a difference, but I'll change it now so that we can see how it works out.

Reply:
I have gone a different route and moved these to a separate header which has to be included
separately. This saves us from including `<ostream>` in `Polyhedron.h`. There's no real need to explicitly instantiate `operator<<` then, since these templates are small and not used very often.